modules/top-level/test.nix

```diff
@@ -38,19 +38,44 @@ in
 
       nativeBuildInputs = [ config.finalPackage ];
 
-      # We need to set HOME because neovim will try to create some files
-      #
-      # Because neovim does not return an exitcode when quitting we need to check if there are
-      # errors on stderr
-      buildPhase = lib.optionalString cfg.runNvim ''
-        mkdir -p .cache/nvim
-
-        output=$(HOME=$(realpath .) nvim -mn --headless "+q" 2>&1 >/dev/null)
-        if [[ -n $output ]]; then
-          echo "ERROR: $output"
-          exit 1
-        fi
-      '';
+      # First check warnings/assertions, then run nvim
+      buildPhase =
+        let
+          inherit (config) warnings;
+          assertions = lib.nixvim.modules.getAssertionMessages config.assertions;
+
+          showErr =
+            name: lines:
+            lib.optionalString (lines != [ ]) ''
+              Unexpected ${name}:
+              ${lib.concatStringsSep "\n" (lib.map (v: "- ${v}") lines)}
+            '';
+
+          errors = lib.foldlAttrs (
+            err: name: lines:
+            err + showErr name lines
+          ) "" { inherit assertions warnings; };
+        in
+        ''
+          ${lib.toShellVars { inherit errors; }}
+          if [ -n "$errors" ]; then
+            echo -n "$errors"
+            exit 1
+          fi
+        ''
```

Comment:
We could move some of this logic into the nix eval by doing something like this instead:

```nix
optionalString (errors != "") ''
  echo -n ${lib.escapeShellArg errors}
  exit 1
''
```

Personal preference, it's fine to do it in the bash build script too.

Reply:
It actually feels kinda weird to take a nix variable and inject it into the bash just to evaluate it again in bash.